app/src/main/java/ca/unb/mobiledev/reflexrevolution/activities/GameActivity.java

```diff
@@ -138,66 +145,44 @@ public void onTick(long l) {
-            //When timer hits zero, start GameOverActivity and pass it the score
+            //When timer hits zero, end game unless instruction requires no input
             public void onFinish() {
-                Intent intent = new Intent(GameActivity.this, GameOverActivity.class);
-                intent.putExtra("Score", score);
-                intent.putExtra("GameMode", gameMode);
-                intent.putExtra("Difficulty", difficulty);
-                startActivity(intent);
-                finish();
+                if(currentInstruction == Instruction.DONTTAP){
+                    detectInput(Instruction.DONTTAP);
+                }
+                else {
+                    endGame();
+                }
             }
         }.start();
     }
 
     //Prepare and start new instruction loop
     private void gameLoop() {
         currentInstruction = getRandomInstruction();
-        displayInstruction();
+        InstructionUtil.displayInstruction(currentInstruction, layout, this);
         registerListeners();
         newTimer();
     }
 
     //Returns a random instruction from instructions
     private Instruction getRandomInstruction() { return instructions.get(rand.nextInt(instructions.size())); }
 
-    //Display UI elements for the current instruction, and set up any necessary input receivers
-    private void displayInstruction() {
-        TextView label;
-        switch (currentInstruction){
-            case BUTTON:
-                Button button = new Button(this);
-                button.setText("PRESS");
-                button.setTextSize(TypedValue.COMPLEX_UNIT_SP, 34);
-                button.setOnClickListener(v -> detectInput(Instruction.BUTTON));
-                layout.addView(button);
-                break;
-
-            case SHAKE:
-                label = new TextView(this);
-                label.setText("SHAKE");
-                label.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
-                label.setTextSize(TypedValue.COMPLEX_UNIT_SP, 34);
-                layout.addView(label);
-                break;
-
-            case JUMP:
-                label = new TextView(this);
-                label.setText("JUMP");
-                label.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
-                label.setTextSize(TypedValue.COMPLEX_UNIT_SP, 34);
-                layout.addView(label);
-                break;
-        }
-    }
-
     //Clear all added UI elements
     private void resetUI() {
         layout.removeAllViews();
     }
 
     //Handle any inputs received
     private void detectInput(Instruction instruction) {
+        //If instruction is don't tap and a tap input is received, end game
+        if(currentInstruction == Instruction.DONTTAP  &&
+                (instruction == Instruction.TAP ||
+                 instruction == Instruction.DOUBLETAP ||
+                 instruction == Instruction.HOLD)) {
+            endGame();
+            return;
+        }
```

Comment:
Here you can just check for `Instruction.TAP` since a `TAP` event is always triggered as part of `DOUBLETAP` and `HOLD`. Like this:
```java
if(currentInstruction == Instruction.DONTTAP && instruction == Instruction.TAP) {...}
```

Reply:
Good point, I didn't consider this. I'll do that.